File: web/src/lib/sessionWorkspace/editSummary.ts
/**
 * Edit-shape detection and diff stats for the timeline (see
 * `docs/specs/transcript-action-visibility.md`).
 *
 * A collapsed row must be able to say *which file changed and how much* without
 * a click, so stats are needed from `formatActivitySummary()` — which runs on
 * every render. `lineDiff()` is O(n*m), so every computation is memoized per
 * interaction and guarded by a cell budget checked *before* the LCS runs.
 */

import type { ToolInteraction } from "./types";
import { lineDiff } from "./diff";

/** Max LCS cells (old lines × new lines) we will spend on one interaction. */
export const DIFF_CELL_BUDGET = 250_000;

export type EditPatch =
  | { kind: "replace"; oldStr: string; newStr: string }
  | { kind: "create"; content: string }
  | { kind: "delete"; content: string }
  | { kind: "patch"; text: string };

export type EditStat = {
  filePath: string | null;
  /** Basename for collapsed headers; full path stays in the expanded body. */
  fileName: string | null;
  added: number;
  removed: number;
  /** False when the shape is unknown or the diff exceeded the budget. */
  hasStat: boolean;
  patch: EditPatch | null;
};

const NO_STAT: EditStat = {
  filePath: null,
  fileName: null,
  added: 0,
  removed: 0,
  hasStat: false,
  patch: null,
};

function str(value: unknown): string | null {
  return typeof value === "string" ? value : null;
}

function basename(path: string | null): string | null {
  if (!path) return null;
  const parts = path.split("/").filter(Boolean);
  return parts.length > 0 ? parts[parts.length - 1] : path;
}

function countLines(text: string): number {
  return text === "" ? 0 : text.split("\n").length;
}

/** Path field names vary across providers (Claude, Codex, OpenCode, Cursor). */
function editFilePath(input: Record<string, unknown>): string | null {
  return (
    str(input.file_path) ??
    str(input.path) ??
    str(input.filePath) ??
    str(input.filename) ??
    null
  );
}

/**
 * Classify an edit-tool input into one of the shapes we can render. Returns
 * null for shapes we do not understand — never fabricate a stat.
 */
export function editPatchFromInput(
  input: Record<string, unknown> | null | undefined,
): EditPatch | null {
  if (!input) return null;

  const oldStr = str(input.old_string) ?? str(input.oldString) ?? str(input.old_str);
  const newStr = str(input.new_string) ?? str(input.newString) ?? str(input.new_str);
  if (oldStr !== null && newStr !== null) return { kind: "replace", oldStr, newStr };
  // `old_string` alone is a removal; `new_string` alone is an insertion.
  if (oldStr !== null) return { kind: "delete", content: oldStr };
  if (newStr !== null) return { kind: "create", content: newStr };

  const patchText = str(input.patch) ?? str(input.diff);
  if (patchText !== null) return { kind: "patch", text: patchText };

  const content = str(input.content) ?? str(input.contents) ?? str(input.text);
  if (content !== null) return { kind: "create", content };

  return null;
}

/**
 * Count `+`/`-` lines in unified-patch text. `+++`/`---` are file headers, not
 * content, and `\ No newline at end of file` is a marker.
 */
function patchStats(text: string): { added: number; removed: number } {
  let added = 0;
  let removed = 0;
  for (const line of text.split("\n")) {
    if (line.startsWith("+++") || line.startsWith("---")) continue;
    if (line.startsWith("+")) added += 1;
    else if (line.startsWith("-")) removed += 1;
  }
  return { added, removed };
}

function computeEditStat(interaction: ToolInteraction): EditStat {
  const presentation = interaction.presentation;
  const raw = interaction.callEvent?.tool_input_json;
  const presented = presentation?.wrapper_recedes ? presentation.tool_input_json : raw;
  const input =
    presented && typeof presented === "object" && !Array.isArray(presented)
      ? (presented as Record<string, unknown>)
      : null;
  if (!input) return NO_STAT;

  const filePath = editFilePath(input);
  const patch = editPatchFromInput(input);
  if (!patch) {
    // Known file, unknown shape: still worth naming the file, without a stat.
    if (!filePath) return NO_STAT;
    return { ...NO_STAT, filePath, fileName: basename(filePath) };
  }

  const base = { filePath, fileName: basename(filePath), patch };

  if (patch.kind === "create") {
    const added = countLines(patch.content);
    return { ...base, added, removed: 0, hasStat: true };
  }
  if (patch.kind === "delete") {
    const removed = countLines(patch.content);
    return { ...base, added: 0, removed, hasStat: true };
  }
  if (patch.kind === "patch") {
    const { added, removed } = patchStats(patch.text);
    return { ...base, added, removed, hasStat: true };
  }

  // Replace: budget check *before* paying for the LCS table.
  const oldLines = countLines(patch.oldStr);
  const newLines = countLines(patch.newStr);
  if (oldLines * newLines > DIFF_CELL_BUDGET) {
    return { ...base, added: 0, removed: 0, hasStat: false };
  }

  let added = 0;
  let removed = 0;
  for (const line of lineDiff(patch.oldStr, patch.newStr)) {
    if (line.kind === "add") added += 1;
    else if (line.kind === "remove") removed += 1;
  }
  return { ...base, added, removed, hasStat: true };
}

const editStatCache = new WeakMap<ToolInteraction, EditStat>();

/** Memoized per interaction — consulted from render paths. */
export function getEditStat(interaction: ToolInteraction): EditStat {
  const cached = editStatCache.get(interaction);
  if (cached) return cached;
  const stat = computeEditStat(interaction);
  editStatCache.set(interaction, stat);
  return stat;
}

/** `timelineModel.ts +4 −1`, `timelineModel.ts` when no stat is available. */
export function formatEditStat(stat: EditStat): string | null {
  const name = stat.fileName;
  if (!name) return null;
  if (!stat.hasStat) return name;
  return `${name} +${stat.added} −${stat.removed}`;
}

File: ios/Sources/Shared/EditSummary.swift
import Foundation

/// Edit-shape detection and diff stats, mirroring
/// `web/src/lib/sessionWorkspace/editSummary.ts`. Both clients must produce
/// byte-identical collapsed summaries against the shared fixture corpus.
///
/// See `docs/specs/transcript-action-visibility.md`.
enum EditSummary {
    /// Max LCS cells (old lines × new lines) spent on one interaction.
    static let diffCellBudget = 250_000

    enum Patch: Equatable {
        case replace(oldStr: String, newStr: String)
        case create(content: String)
        case delete(content: String)
        case patch(text: String)
    }

    struct Stat: Equatable {
        var filePath: String?
        /// Basename for collapsed headers; the full path stays in the body.
        var fileName: String?
        var added: Int = 0
        var removed: Int = 0
        /// False when the shape is unknown or the diff exceeded the budget.
        var hasStat: Bool = false
        var patch: Patch?

        static let none = Stat()
    }

    private static func countLines(_ text: String) -> Int {
        text.isEmpty ? 0 : text.components(separatedBy: "\n").count
    }

    private static func basename(_ path: String?) -> String? {
        guard let path else { return nil }
        let parts = path.split(separator: "/").map(String.init)
        return parts.last ?? path
    }

    /// Path field names vary across providers (Claude, Codex, OpenCode, Cursor).
    private static func filePath(_ event: SessionEvent) -> String? {
        for key in ["file_path", "path", "filePath", "filename"] {
            if let value = TimelineBuilder.presentedToolInputString(event, key), !value.isEmpty {
                return value
            }
        }
        return nil
    }

    /// Classify an edit-tool input into a shape we can render. Returns nil for
    /// shapes we do not understand — never fabricate a stat.
    static func patch(for event: SessionEvent) -> Patch? {
        func input(_ keys: [String]) -> String? {
            for key in keys {
                if let value = TimelineBuilder.presentedToolInputString(event, key) { return value }
            }
            return nil
        }

        let oldStr = input(["old_string", "oldString", "old_str"])
        let newStr = input(["new_string", "newString", "new_str"])
        if let oldStr, let newStr { return .replace(oldStr: oldStr, newStr: newStr) }
        // `old_string` alone is a removal; `new_string` alone is an insertion.
        if let oldStr { return .delete(content: oldStr) }
        if let newStr { return .create(content: newStr) }

        if let patchText = input(["patch", "diff"]) { return .patch(text: patchText) }
        if let content = input(["content", "contents", "text"]) { return .create(content: content) }
        return nil
    }

    /// Count `+`/`-` lines in unified-patch text. `+++`/`---` are file headers.
    private static func patchStats(_ text: String) -> (added: Int, removed: Int) {
        var added = 0
        var removed = 0
        for line in text.components(separatedBy: "\n") {
            if line.hasPrefix("+++") || line.hasPrefix("---") { continue }
            if line.hasPrefix("+") { added += 1 } else if line.hasPrefix("-") { removed += 1 }
        }
        return (added, removed)
    }

    /// No cache here: unlike the web render loop, `activitySummary` is not
    /// called per frame, and a `static var` cache would be a concurrency hazard.
    /// The budget check below is what keeps the cost bounded.
    static func stat(for event: SessionEvent) -> Stat {
        let path = filePath(event)
        guard let patch = patch(for: event) else {
            // Known file, unknown shape: still name the file, without a stat.
            guard let path else { return .none }
            return Stat(filePath: path, fileName: basename(path))
        }

        var stat = Stat(filePath: path, fileName: basename(path), patch: patch)
        switch patch {
        case .create(let content):
            stat.added = countLines(content)
            stat.hasStat = true
        case .delete(let content):
            stat.removed = countLines(content)
            stat.hasStat = true
        case .patch(let text):
            let counts = patchStats(text)
            stat.added = counts.added
            stat.removed = counts.removed
            stat.hasStat = true
        case .replace(let oldStr, let newStr):
            // Budget check *before* paying for the LCS table.
            guard countLines(oldStr) * countLines(newStr) <= diffCellBudget else { return stat }
            // Count from the diff we already have to build, rather than running
            // a second LCS pass just for the totals.
            for line in lineDiff(oldStr: oldStr, newStr: newStr) {
                if line.kind == .add { stat.added += 1 } else if line.kind == .remove { stat.removed += 1 }
            }
            stat.hasStat = true
        }
        return stat
    }

    struct DiffLine: Equatable {
        enum Kind: String { case equal, add, remove }
        let kind: Kind
        let text: String
    }

    /// Renderable diff lines for a stat's patch, mirroring `EditDiffView` on
    /// web. Returns nil for shapes with no patch.
    static func diffLines(for stat: Stat, context: Int = 2) -> [DiffLine]? {
        guard let patch = stat.patch else { return nil }
        switch patch {
        case .create(let content):
            return content.components(separatedBy: "\n").map { DiffLine(kind: .add, text: $0) }
        case .delete(let content):
            return content.components(separatedBy: "\n").map { DiffLine(kind: .remove, text: $0) }
        case .patch(let text):
            return text.components(separatedBy: "\n").map { line in
                if line.hasPrefix("+"), !line.hasPrefix("+++") { return DiffLine(kind: .add, text: line) }
                if line.hasPrefix("-"), !line.hasPrefix("---") { return DiffLine(kind: .remove, text: line) }
                return DiffLine(kind: .equal, text: line)
            }
        case .replace(let oldStr, let newStr):
            guard stat.hasStat else { return nil }
            return collapseUnchanged(lineDiff(oldStr: oldStr, newStr: newStr), context: context)
        }
    }

    /// LCS line diff. Scoped to the small strings edit tools produce; callers
    /// must apply `diffCellBudget` first.
    static func lineDiff(oldStr: String, newStr: String) -> [DiffLine] {
        let a = oldStr.isEmpty ? [] : oldStr.components(separatedBy: "\n")
        let b = newStr.isEmpty ? [] : newStr.components(separatedBy: "\n")
        let m = a.count
        let n = b.count
        if m == 0 { return b.map { DiffLine(kind: .add, text: $0) } }
        if n == 0 { return a.map { DiffLine(kind: .remove, text: $0) } }

        var dp = [[Int]](repeating: [Int](repeating: 0, count: n + 1), count: m + 1)
        for i in stride(from: m - 1, through: 0, by: -1) {
            for j in stride(from: n - 1, through: 0, by: -1) {
                dp[i][j] = a[i] == b[j] ? dp[i + 1][j + 1] + 1 : max(dp[i + 1][j], dp[i][j + 1])
            }
        }

        var out: [DiffLine] = []
        var i = 0
        var j = 0
        while i < m, j < n {
            if a[i] == b[j] {
                out.append(DiffLine(kind: .equal, text: a[i]))
                i += 1
                j += 1
            } else if dp[i + 1][j] >= dp[i][j + 1] {
                out.append(DiffLine(kind: .remove, text: a[i]))
                i += 1
            } else {
                out.append(DiffLine(kind: .add, text: b[j]))
                j += 1
            }
        }
        while i < m { out.append(DiffLine(kind: .remove, text: a[i])); i += 1 }
        while j < n { out.append(DiffLine(kind: .add, text: b[j])); j += 1 }
        return out
    }

    /// Collapse long unchanged runs into a single marker, keeping `context`
    /// lines either side of each change.
    static func collapseUnchanged(_ lines: [DiffLine], context: Int = 2) -> [DiffLine] {
        guard lines.count > context * 2 + 1 else { return lines }
        var keep = [Bool](repeating: false, count: lines.count)
        for (i, line) in lines.enumerated() where line.kind != .equal {
            for k in max(0, i - context)...min(lines.count - 1, i + context) { keep[k] = true }
        }

        var out: [DiffLine] = []
        var skipped = 0
        func flush() {
            guard skipped > 0 else { return }
            out.append(DiffLine(
                kind: .equal,
                text: "… \(skipped) unchanged line\(skipped == 1 ? "" : "s") …"
            ))
            skipped = 0
        }
        for (i, line) in lines.enumerated() {
            if keep[i] {
                flush()
                out.append(line)
            } else {
                skipped += 1
            }
        }
        flush()
        return out
    }

    /// `timelineModel.ts +4 −1`, or the bare name when no stat is available.
    static func format(_ stat: Stat) -> String? {
        guard let name = stat.fileName else { return nil }
        guard stat.hasStat else { return name }
        return "\(name) +\(stat.added) −\(stat.removed)"
    }
}
